rfcs/convergence/as-prop.md

```diff
@@ -0,0 +1,197 @@
+
+```tsx
+<Button as={Link} /> // renders <Link /> + additional props
+```
+
+These usages are nightmare for typings as we must use really complicated generics ([`ComponentWithAs`](https://github.com/microsoft/fluentui/blob/c27c8fa7ae589a61267bd0a6f0f1c7a49a6be9c2/packages/react-compose/src/types.ts#L21)) to support polymorphism. There are few similar implementations:
+
+- [react-polymorphic-box](https://github.com/kripod/react-polymorphic-box)
+- [forwardRefWithAs from Reach UI](https://github.com/reach/reach-ui/blob/e1a6fba7042e1e023c4a3a4e7c15a13225500667/packages/utils/src/index.tsx#L191-L199)
+
+Anyway, with these generics our component can still throw on unsupported props:
+
+```tsx
+// ✅ TS compiler will throw an error because "someProp" is not supported by "Button" and "Link"
+<Button as={Link} someProp />
+```
+
+But, it still creates problems for customers when they will decide to compose our components:
+
+```tsx
+// Uses a React HOC to control rendering
+// https://reactjs.org/docs/react-api.html#reactmemo
+const MemoButton = React.memo(Button);
+
+// ⚠ It should be enough for customers to use "ButtonProps", but "ComponentWithAs" requires different typings
+function PrimaryButton(props: ButtonProps) {
+  return <Button {...props} primary />;
+}
+
+function App() {
+  return (
+    <>
+      {/* ✅ TS throws as "href" is not supported for "div" elements */}
+      <Button as="div" href="#" />
+      {/* no errors 🤷‍♂️ */}
+      <MemoButton as="div" href="#" />
+
+      {/* 💣 "as" and appropriate props are created by "ComponentWithAs" generic */}
+      {/*     generic thus are not inside "ButtonProps" */}
+      <PrimaryButton as="div" />
+      <PrimaryButton type="submit" />
+    </>
+  );
+}
+```
+
+**Retrospective note:** It was required to support React elements in `as` prop for [Semantic UI React](https://github.com/Semantic-Org/Semantic-UI-React) due elements composition:
+
+```tsx
+<>
+  {/* ✅ will render a proper markup expected by CSS styling */}
+  {/* <div class="ui sidebar menu"/> */}
+  <Sidebar as={Menu} />
+
+  {/* ❌ styles will be different in this case */}
+  {/* <div class="ui sidebar"><div class="ui menu" /></div> */}
+  <Sidebar>
+    <Menu />
+  </Sidebar>
+</>
+```
+
+On Fluent side we don't have such restrictions from styling system.
+
+## Event handlers problem
+
+_It's a real scenario from Teams._
+
+Without this change we know that in `as` we can get a memoized component that forces us memoize event handlers:
+
+```tsx
+function Button() {
+  // 💣 Oops, this will break memoized input, should be wrapped with React.useCallback()
+  const handleClick = () => {};
+
+  return <Element onClick={handleClick} />;
+}
+
+<Button as={MemoEl} />;
+```
+
+This can be extremely tricky for us and customers with hooks approach as they may not be aware of these requirements:
```

Comment:
Not entirely sure how relevant the below example is to `as` prop, seems more like an issue with a hook approach tbh